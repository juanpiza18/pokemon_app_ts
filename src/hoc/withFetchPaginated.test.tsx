import React from "react";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import WithFetchPaginated from "./withFetchPaginated.hoc";
import PokemonContext from "../context/pokemonContext";

// Componente dummy para el HoC
const DummyComponent: React.FC<{ foo: string }> = ({ foo }) => {
  return <div className="dummy__component">{foo}</div>;
};

const fetchDataMock = jest.fn();
const paginationNextMock = jest.fn();
const paginationPrevMock = jest.fn();
const mockFilterAllList = jest.fn();

const DummyContext = React.createContext({
  loading: false,
  dataList: [],
  filteredList: [],
  filterAllList: mockFilterAllList,
  fetchData: fetchDataMock,
  paginationNext: paginationNextMock,
  paginationPrev: paginationPrevMock,
});

const MockWithFetchComponent = WithFetchPaginated(DummyComponent, DummyContext);

describe("WithFetchPaginated", () => {
  it("should render With Fetch Paginated and Match snapshot", () => {
    const { container } = render(<MockWithFetchComponent />);
    expect(screen.getByTestId("list")).toBeTruthy();
    expect(screen.getByTestId("pagination")).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  it("should call fetchData on mount", () => {
    render(<MockWithFetchComponent />);
    expect(fetchDataMock).toHaveBeenCalledTimes(1);
  });

  it("should call pagination prve on click", () => {
    const { container } = render(<MockWithFetchComponent />);
    fireEvent.click(container.querySelectorAll("button")[0]);
    expect(paginationPrevMock).toHaveBeenCalledTimes(1);
  });

  it("should call pagination next on click", () => {
    const { container } = render(<MockWithFetchComponent />);
    fireEvent.click(container.querySelectorAll("button")[1]);
    expect(paginationNextMock).toHaveBeenCalledTimes(1);
  });

  it("should filter data when filter is not null", async () => {
    await waitFor(() => {
      render(<MockWithFetchComponent filter="Bulbasaur" />);
    });
    expect(fetchDataMock).not.toHaveBeenCalled();
    expect(mockFilterAllList).toHaveBeenCalledWith("Bulbasaur");
  });
});
