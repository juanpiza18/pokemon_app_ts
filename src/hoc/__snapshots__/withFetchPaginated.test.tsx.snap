// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WithFetchPaginated should render With Fetch Paginated and Match snapshot 1`] = `
<div>
  <div
    class="flex__container"
  >
    <div
      class="list__actions"
      data-testid="pagination"
    >
      <button
        class="btn"
      >
        ← Prev
      </button>
      <button
        class="btn"
      >
        Next →
      </button>
    </div>
    <div
      class="card__grid"
      data-testid="list"
    >
      <div
        class="dummy__component"
      />
    </div>
  </div>
</div>
`;
